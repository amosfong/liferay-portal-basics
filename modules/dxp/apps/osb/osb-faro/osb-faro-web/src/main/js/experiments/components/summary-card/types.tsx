export enum Status {
	Completed = 'completed',
	Draft = 'draft',
	FinishedNoWinner = 'finished_no_winner',
	FinishedWinner = 'finished_winner',
	Running = 'running',
	Scheduled = 'scheduled',
	Terminated = 'terminated'
}
