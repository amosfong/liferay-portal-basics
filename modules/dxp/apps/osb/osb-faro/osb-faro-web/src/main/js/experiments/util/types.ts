export type FormatYAxisFn = (metricUnit: string) => (value: number) => string;

export type GetFormattedMedianFn = (
	median: number,
	metric: MetricName
) => string;

export type GetMetricNameFn = (metric: MetricName) => string;

export type GetMetricUnitFn = (metric: MetricName) => string;

export type GetStatusColorFn = (status: Status) => string;

export type GetStatusNameFn = (status: Status) => string;

export type GetShortIntervals = (intervals: Array<Date>) => Array<Date>;

export type GetTicksFn = (maxValue: number) => Array<number>;

export type GetVariantLabels = (experiment: {
	bestVariant?: Variant;
	dxpVariantId: string;
	publishedDXPVariantId?: string;
	status: Status;
	winnerDXPVariantId?: string;
}) => Array<{status: string; value: string}>;

type MetricName =
	| 'BOUNCE_RATE'
	| 'CLICK_RATE'
	| 'MAX_SCROLL_DEPTH'
	| 'TIME_ON_PAGE';

type MergedVariant = {
	changes: number;
	confidenceInterval: Array<number>;
	control: boolean;
	dxpVariantId: string;
	dxpVariantName: string;
	improvement: number;
	median: number;
	probabilityToWin: number;
	trafficSplit: number;
	uniqueVisitors: number;
};

export type MergedVariantsFn = (
	variants: Array<Variant>,
	variantMetrics: Array<VariantMetric>
) => Array<MergedVariant>;

type Status =
	| 'COMPLETED'
	| 'DRAFT'
	| 'FINISHED_NO_WINNER'
	| 'FINISHED_WINNER'
	| 'PAUSED'
	| 'RUNNING'
	| 'SCHEDULED'
	| 'TERMINATED';

type Variant = {
	changes: number;
	control: boolean;
	dxpVariantId: string;
	dxpVariantName: string;
	trafficSplit: number;
	uniqueVisitors: number;
};

type VariantMetric = {
	confidenceInterval: Array<number>;
	dxpVariantId: string;
	improvement: number;
	median: number;
	probabilityToWin: number;
};
