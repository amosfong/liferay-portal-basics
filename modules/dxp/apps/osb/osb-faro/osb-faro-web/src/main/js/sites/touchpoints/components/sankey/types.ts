export enum Type {
	Previous = 'previous',
	Following = 'following'
}

export enum TitleKey {
	Direct = 'direct',
	DropOffs = 'drop-offs',
	Others = 'others'
}
