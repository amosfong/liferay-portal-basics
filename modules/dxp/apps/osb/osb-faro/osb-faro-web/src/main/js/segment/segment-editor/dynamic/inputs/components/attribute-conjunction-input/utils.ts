import {BetweenNumber} from '../BetweenNumberInput';
import {BOOLEAN_LABELS_MAP} from 'event-analysis/utils/utils';
import {DataTypes, Operators} from 'event-analysis/utils/types';
import {DateRange} from 'shared/components/DateRangeInput';
import {
	FunctionalOperators,
	NotOperators,
	RelationalOperators,
	STRING_OPTIONS
} from '../../../utils/constants';
import {isNumber} from 'lodash';
import {isValid} from '../../../utils/utils';

const ATTRIBUTES_DATE_AND_DURATION_OPERATORS_LONGHAND_LABELS_MAP = {
	[Operators.EQ]: Liferay.Language.get('is-fragment'),
	[Operators.GT]: Liferay.Language.get('is-after-fragment'),
	[Operators.LT]: Liferay.Language.get('is-before-fragment')
};

const ATTRIBUTES_DATE_AND_DURATION_OPTIONS = [
	Operators.LT,
	Operators.EQ,
	Operators.GT
];

export const ATTRIBUTES_NUMBER_OPERATOR_LONGHAND_LABELS_MAP = {
	[Operators.EQ]: Liferay.Language.get('is-equal-to-fragment'),
	[Operators.GT]: Liferay.Language.get('greater-than-fragment'),
	[Operators.LT]: Liferay.Language.get('less-than-fragment'),
	[Operators.NE]: Liferay.Language.get('is-not-equal-to-fragment')
};

const ATTRIBUTE_NUMBER_OPTIONS = [
	Operators.EQ,
	Operators.GT,
	Operators.LT,
	Operators.NE
];

const ATTRIBUTES_STRING_OPERATOR_LABELS_MAP = {
	[FunctionalOperators.Contains]: Liferay.Language.get('contains-fragment'),
	[NotOperators.NotContains]: Liferay.Language.get(
		'does-not-contain-fragment'
	),
	[RelationalOperators.EQ]: Liferay.Language.get('is-fragment'),
	[RelationalOperators.NE]: Liferay.Language.get('is-not-fragment')
};

export const createOption = (option, dataType: DataTypes) => {
	const LABELS_MAP = {
		[DataTypes.Boolean]: BOOLEAN_LABELS_MAP,
		[DataTypes.Date]: ATTRIBUTES_DATE_AND_DURATION_OPERATORS_LONGHAND_LABELS_MAP,
		[DataTypes.Duration]: ATTRIBUTES_DATE_AND_DURATION_OPERATORS_LONGHAND_LABELS_MAP,
		[DataTypes.Number]: ATTRIBUTES_NUMBER_OPERATOR_LONGHAND_LABELS_MAP,
		[DataTypes.String]: ATTRIBUTES_STRING_OPERATOR_LABELS_MAP // "NotContains" differs from segment-editor and event-analysis. We should be able to use the evente-analysis version once we move away from odata.
	};

	return {
		label: LABELS_MAP[dataType][option],
		value: option
	};
};

export const getOperatorOptions = (dataType: DataTypes) => {
	const OPERATOR_OPTIONS = {
		[DataTypes.Date]: ATTRIBUTES_DATE_AND_DURATION_OPTIONS,
		[DataTypes.Duration]: ATTRIBUTES_DATE_AND_DURATION_OPTIONS,
		[DataTypes.Number]: ATTRIBUTE_NUMBER_OPTIONS,
		[DataTypes.String]: STRING_OPTIONS // STRING_OPTIONS is provided from the segment-editor utils as "NotContains" differs from segment-editor and event-analysis. We should be able to use the evente-analysis version once we move away from odata.
	};

	return OPERATOR_OPTIONS[dataType]?.map(option =>
		createOption(option, dataType)
	);
};

export const getDefaultAttributeOperator = (
	dataType: DataTypes
): RelationalOperators | FunctionalOperators => {
	switch (dataType) {
		case DataTypes.Boolean:
		case DataTypes.Date:
			return RelationalOperators.EQ;
		case DataTypes.Duration:
		case DataTypes.Number:
			return RelationalOperators.GT;
		case DataTypes.String:
		default:
			return FunctionalOperators.Contains;
	}
};

export const getDefaultAttributeValue = (
	dataType: DataTypes,
	operatorName: RelationalOperators | FunctionalOperators
): string | {end: number | string; start: number | string} => {
	if (
		operatorName === FunctionalOperators.Between &&
		[DataTypes.Number, DataTypes.Date].filter(type => type === dataType)
	) {
		return {end: '', start: ''};
	}

	switch (dataType) {
		case DataTypes.Boolean:
			return 'true';
		case DataTypes.Date:
		case DataTypes.Number:
		case DataTypes.Duration:
		case DataTypes.String:
		default:
			return '';
	}
};

export const validateAttributeValue = (
	value: string | number | BetweenNumber | DateRange,
	dataType: DataTypes,
	operatorName?: FunctionalOperators | RelationalOperators
): boolean => {
	if (
		operatorName === FunctionalOperators.Between &&
		[DataTypes.Number, DataTypes.Date].filter(type => type === dataType)
	) {
		const {end, start} = value as BetweenNumber;

		return isValid(end) && isValid(start);
	}

	switch (dataType) {
		case DataTypes.Boolean:
			return value === 'true' || value === 'false';
		case DataTypes.Duration:
		case DataTypes.Number:
			return isNumber(value);
		case DataTypes.Date:
		case DataTypes.String:
		default:
			return isValid(value);
	}
};
