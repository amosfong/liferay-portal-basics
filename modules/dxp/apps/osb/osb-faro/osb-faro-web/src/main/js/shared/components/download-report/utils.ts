import moment from 'moment';
import {buildOrderByFields} from 'shared/util/pagination';
import {DEFAULT_DATE_FORMAT} from 'shared/util/date';
import {DEFAULT_RANGE_SELECTORS} from 'shared/hooks/useQueryRangeSelectors';
import {INDIVIDUALS} from 'shared/util/router';
import {RangeKeyTimeRanges} from 'shared/util/constants';
import {RangeSelectors} from 'shared/types';
import {useParams} from 'react-router-dom';

export function formatDate(date: string | Date) {
	return moment(date).format(DEFAULT_DATE_FORMAT);
}

export enum CSVType {
	Blog = 'blog',
	Document = 'document',
	Event = 'event',
	Form = 'form',
	Individual = 'individual',
	Journal = 'journal',
	Membership = 'membership',
	Page = 'page'
}

export function useDownloadCSV({
	assetId,
	assetType,
	individualId,
	segmentId,
	type
}: {
	assetId?: string;
	assetType?: string;
	individualId?: string;
	segmentId?: string;
	type: CSVType;
}) {
	const {channelId, groupId, title} = useParams();

	return (rangeSelectors: RangeSelectors = DEFAULT_RANGE_SELECTORS) => {
		const searchParams = new URLSearchParams(location.search);

		const field = searchParams.get('field');
		const query = searchParams.get('query');
		const sortOrder = searchParams.get('sortOrder');

		let url = `/o/faro/main/${groupId}/reports/export/csv/${type}?channelId=${channelId}`;

		if (rangeSelectors.rangeKey === RangeKeyTimeRanges.CustomRange) {
			url += '&rangeKey=CUSTOM';
			url += `&fromDate=${formatDate(rangeSelectors?.rangeStart)}`;
			url += `&toDate=${formatDate(rangeSelectors?.rangeEnd)}`;
		} else {
			url += `&rangeKey=${rangeSelectors.rangeKey}`;
		}

		const optionalParams = {
			assetId: assetId && encodeURIComponent(assetId),
			assetTitle: title,
			assetType,
			individualId,
			orderByFields:
				field && sortOrder
					? encodeURIComponent(
							JSON.stringify(
								buildOrderByFields(
									{field, sortOrder},
									INDIVIDUALS
								)
							)
					  )
					: null,
			query,
			segmentId
		};

		Object.entries(optionalParams).forEach(([key, value]) => {
			if (value) {
				url += `&${key}=${value}`;
			}
		});

		return url;
	};
}

export const MAX_CSV_ENTRIES = 10000;
