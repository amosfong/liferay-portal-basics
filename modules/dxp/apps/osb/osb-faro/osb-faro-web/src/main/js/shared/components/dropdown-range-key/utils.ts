import moment, {Moment} from 'moment';
import momentTimezone from 'moment-timezone';
import {getDate} from 'shared/util/date';
import {RangeKeyTimeRanges} from 'shared/util/constants';
import {TIME_RANGE_LABELS} from 'shared/util/constants';
import {utcFormat} from 'd3';

type TimeRange = {
	description: string;
	value: RangeKeyTimeRanges;
	label: string;
};

export function formatDateWithTimezone(
	timeZoneId: string,
	date = momentTimezone.utc()
): Moment {
	const currentDateWithTimeZone = momentTimezone.tz(date, timeZoneId);
	const hours = currentDateWithTimeZone.utcOffset() / 60;
	const signal = hours < 0 ? 'subtract' : 'add';

	return date.clone()[signal](Math.abs(hours), 'hours');
}

export function formatDateRange(date, rangeKey) {
	if (
		`${rangeKey}` === RangeKeyTimeRanges.Last24Hours ||
		`${rangeKey}` === RangeKeyTimeRanges.Yesterday
	) {
		return utcFormat('%d %b, %I %p')(date);
	}

	return utcFormat('%d %b')(date);
}

export function formatTimeRange(timeRange) {
	return timeRange
		.map(({endDate, rangeKey, startDate}) => {
			const start = formatDateRange(getDate(startDate), rangeKey);
			const end = formatDateRange(getDate(endDate), rangeKey);
			const label = `${start} - ${end}`;

			return {
				description: startDate && endDate && label,
				label: TIME_RANGE_LABELS[rangeKey],
				value: `${rangeKey}` as RangeKeyTimeRanges
			};
		})
		.sort((a, b) => parseInt(a.value) - parseInt(b.value));
}

const filterItemsByRetention = (
	timeRange: Array<TimeRange>,
	retentionPeriod: number
): Array<TimeRange> =>
	timeRange.filter(
		item =>
			!(
				retentionPeriod === 7 &&
				item.value === RangeKeyTimeRanges.LastYear
			)
	);

const filterLegacyItems = (
	timeRange: Array<TimeRange>,
	retentionPeriod: number
): Array<TimeRange> =>
	filterItemsByRetention(
		timeRange.filter(({value}) =>
			[
				RangeKeyTimeRanges.Last24Hours,
				RangeKeyTimeRanges.Yesterday,
				RangeKeyTimeRanges.Last7Days,
				RangeKeyTimeRanges.Last28Days,
				RangeKeyTimeRanges.Last30Days,
				RangeKeyTimeRanges.Last90Days
			].includes(value)
		),
		retentionPeriod
	);

const filterMoreItems = (
	timeRange: Array<TimeRange>,
	rangeKeys: Array<RangeKeyTimeRanges>,
	retentionPeriod: number
): Array<TimeRange> =>
	filterItemsByRetention(
		timeRange.filter(
			item =>
				!rangeKeys
					.filter(value => !rangeKeys.includes(value))
					.includes(item.value)
		),
		retentionPeriod
	);

const filterItems = (
	timeRange: Array<TimeRange>,
	rangeKey: RangeKeyTimeRanges,
	rangeKeys: Array<RangeKeyTimeRanges>,
	retentionPeriod: number
) =>
	filterItemsByRetention(
		timeRange.filter(
			({value}) => value === rangeKey || rangeKeys.includes(value)
		),
		retentionPeriod
	);

type GetFilteredItems = (props: {
	legacy: boolean;
	rangeKey: RangeKeyTimeRanges;
	rangeKeys?: Array<RangeKeyTimeRanges>;
	retentionPeriod: number;
	seeMore: boolean;
	timeRange: Array<TimeRange>;
}) => Array<TimeRange>;

export const getFilteredItems: GetFilteredItems = ({
	legacy,
	rangeKey,
	rangeKeys = [
		RangeKeyTimeRanges.Last24Hours,
		RangeKeyTimeRanges.Last7Days,
		RangeKeyTimeRanges.Last30Days,
		RangeKeyTimeRanges.Last90Days
	],
	retentionPeriod,
	seeMore,
	timeRange
}) => {
	if (legacy) {
		return filterLegacyItems(timeRange, retentionPeriod);
	}

	if (seeMore) {
		return filterMoreItems(timeRange, rangeKeys, retentionPeriod);
	}

	return filterItems(timeRange, rangeKey, rangeKeys, retentionPeriod);
};

export function getSelectedItem({rangeEnd, rangeKey, rangeStart, timeRange}) {
	if (rangeKey === 'CUSTOM') {
		return {
			label: `${moment(rangeStart).format('ll')} - ${moment(
				rangeEnd
			).format('ll')}`,
			value: 'CUSTOM'
		};
	}

	return timeRange.find(({value}) => value === rangeKey) || timeRange[0];
}
