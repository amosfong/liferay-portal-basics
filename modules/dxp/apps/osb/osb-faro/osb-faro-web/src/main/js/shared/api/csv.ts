import sendRequest from 'shared/util/request';
import {CSVType} from 'shared/components/download-report/utils';

export const fetchCSV = (url: string) => fetch(url);

export const fetchCount = ({
	groupId,
	type,
	...data
}: {
	assetId?: string;
	assetType?: string;
	channelId: string;
	fromDate?: string;
	groupId: string;
	individualId?: string;
	rangeKey?: string;
	toDate?: string;
	type: CSVType;
}) =>
	sendRequest({
		data,
		method: 'GET',
		path: `main/${groupId}/reports/export/csv/${type}/count`
	});
