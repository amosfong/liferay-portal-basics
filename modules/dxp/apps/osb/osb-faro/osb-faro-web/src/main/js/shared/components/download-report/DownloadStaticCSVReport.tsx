import * as API from 'shared/api';
import ClayButton from '@clayui/button';
import ClayForm from '@clayui/form';
import ClayModal, {useModal} from '@clayui/modal';
import React from 'react';
import {addAlert} from 'shared/actions/alerts';
import {Alert} from 'shared/types';
import {CSVType, MAX_CSV_ENTRIES, useDownloadCSV} from './utils';
import {DownloadReportButton} from './DownloadReportButton';
import {sub} from 'shared/util/lang';
import {toLocale} from 'shared/util/numbers';
import {useDispatch} from 'react-redux';
import {useParams} from 'react-router-dom';

interface IDownloadStaticCSVReport {
	disabled: boolean;
	type: CSVType;
	typeLang: string;
	segmentId?: string;
}

export const DownloadStaticCSVReport: React.FC<IDownloadStaticCSVReport> = ({
	disabled,
	segmentId,
	type,
	typeLang
}) => {
	const dispatch = useDispatch();
	const generateURL = useDownloadCSV({segmentId, type});
	const {observer, onOpenChange, open} = useModal();
	const {channelId, groupId} = useParams();

	return (
		<>
			<DownloadReportButton
				disabled={disabled}
				onClick={() => onOpenChange(true)}
			/>

			{open && (
				<Modal
					observer={observer}
					onClose={() => onOpenChange(false)}
					onSubmit={async () => {
						onOpenChange(false);

						try {
							const url = generateURL();
							const response = await API.csv.fetchCSV(url);

							if (!response.ok) {
								throw new Error();
							}

							dispatch(
								addAlert({
									alertType: Alert.Types.Default,
									message: sub(
										Liferay.Language.get(
											'the-x-file-is-being-generated-and-your-download-will-start-soon'
										),
										['CSV']
									) as string
								})
							);

							const a = document.createElement('a');

							a.href = url;
							a.click();

							const count = await API.csv.fetchCount({
								channelId,
								groupId,
								type: CSVType.Individual
							});

							if (count > MAX_CSV_ENTRIES) {
								dispatch(
									addAlert({
										alertType: Alert.Types.Warning,
										message: sub(
											Liferay.Language.get(
												'the-csv-file-reached-x-entries'
											),
											[toLocale(MAX_CSV_ENTRIES)]
										)
									})
								);
							}
						} catch (e) {
							dispatch(
								addAlert({
									alertType: Alert.Types.Error,
									message: Liferay.Language.get(
										'it-was-not-possible-to-generate-a-csv-file-at-this-moment.-please-try-again-later'
									)
								})
							);
						}
					}}
					typeLang={typeLang}
				/>
			)}
		</>
	);
};

const Modal = ({observer, onClose, onSubmit, typeLang}) => (
	<ClayModal observer={observer}>
		<ClayForm
			onSubmit={event => {
				event.preventDefault();

				onSubmit();
			}}
		>
			<ClayModal.Header>
				{Liferay.Language.get('download-reports')}
			</ClayModal.Header>

			<ClayModal.Body>
				<p>
					{
						sub(
							Liferay.Language.get(
								'the-generated-csv-file-supports-up-to-x-entries-per-export-and-it-will-respect-the-current-ordering-and-search-results.-please-ensure-that-any-desired-changes-have-been-successfully-applied-before-downloading-the-x-list'
							),
							[toLocale(MAX_CSV_ENTRIES), typeLang]
						) as string
					}
				</p>
			</ClayModal.Body>

			<ClayModal.Footer
				last={
					<ClayButton.Group spaced>
						<ClayButton
							data-testid='cancel'
							displayType='secondary'
							onClick={onClose}
						>
							{Liferay.Language.get('cancel')}
						</ClayButton>

						<ClayButton data-testid='submit' type='submit'>
							{Liferay.Language.get('download')}
						</ClayButton>
					</ClayButton.Group>
				}
			/>
		</ClayForm>
	</ClayModal>
);
