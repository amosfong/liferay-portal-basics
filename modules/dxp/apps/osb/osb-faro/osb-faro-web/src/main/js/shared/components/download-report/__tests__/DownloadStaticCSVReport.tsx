import * as API from 'shared/api';
import mockStore from 'test/mock-store';
import React from 'react';
import ReactDOM from 'react-dom';
import {addAlert} from 'shared/actions/alerts';
import {Alert} from 'shared/types';
import {BrowserRouter} from 'react-router-dom';
import {cleanup, fireEvent, render, waitFor} from '@testing-library/react';
import {CSVType} from '../utils';
import {DownloadStaticCSVReport} from '../DownloadStaticCSVReport';
import {Provider} from 'react-redux';

jest.unmock('react-dom');

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useParams: () => ({
		channelId: '123',
		groupId: '456'
	})
}));

jest.mock('shared/actions/alerts', () => ({
	actionTypes: {},
	addAlert: jest.fn(() => ({
		meta: {},
		payload: {},
		type: 'addAlert'
	}))
}));

// Mock a.click on DownloadStaticCSVReport to avoid
// break navigation during running unit tests with jest

jest.spyOn(HTMLAnchorElement.prototype, 'click').mockImplementation(() => {});

const WrapperComponent = () => (
	<Provider store={mockStore()}>
		<BrowserRouter>
			<DownloadStaticCSVReport
				disabled={false}
				type={CSVType.Individual}
				typeLang={Liferay.Language.get('individuals')}
			/>
		</BrowserRouter>
	</Provider>
);

describe('DownloadStaticCSVReport', () => {
	afterEach(() => {
		jest.clearAllMocks();
		jest.clearAllTimers();
		cleanup();
	});

	beforeAll(() => {
		jest.useFakeTimers();

		// @ts-ignore

		ReactDOM.createPortal = jest.fn(element => element);
	});

	afterAll(() => {
		jest.useRealTimers();
	});

	it('renders component', () => {
		const {container, getByRole} = render(<WrapperComponent />);

		fireEvent.click(
			getByRole('button', {
				name: /download report/i
			})
		);

		jest.runAllTimers();

		expect(container).toMatchSnapshot();
	});

	it('displays modal content', () => {
		// @ts-ignore

		API.csv.fetchCSV.mockReturnValueOnce(Promise.resolve({ok: true}));

		// @ts-ignore

		API.csv.fetchCount.mockReturnValueOnce(Promise.resolve(9000));

		const {getByRole, queryByTestId, queryByText} = render(
			<WrapperComponent />
		);

		fireEvent.click(
			getByRole('button', {
				name: /download report/i
			})
		);

		jest.runAllTimers();

		expect(
			queryByText(
				'The generated CSV file supports up to 10,000 entries per export and it will respect the current ordering and search results. Please ensure that any desired changes have been successfully applied before downloading the Individuals list.'
			)
		).toBeTruthy();

		expect(queryByTestId('cancel')).toBeTruthy();
		expect(queryByTestId('submit')).toBeTruthy();

		fireEvent.click(queryByTestId('submit'));

		jest.runAllTimers();

		expect(
			queryByText(
				'The generated CSV file supports up to 10,000 entries per export and it will respect the current ordering and search results. Please ensure that any desired changes have been successfully applied before downloading the Individuals list.'
			)
		).toBeFalsy();

		expect(queryByTestId('cancel')).toBeFalsy();
		expect(queryByTestId('submit')).toBeFalsy();
	});

	it('displays info alert about download CSV report', async () => {
		// @ts-ignore

		API.csv.fetchCSV.mockReturnValueOnce(Promise.resolve({ok: true}));

		// @ts-ignore

		API.csv.fetchCount.mockReturnValueOnce(Promise.resolve(9000));

		const {getByRole, queryByTestId} = render(<WrapperComponent />);

		fireEvent.click(
			getByRole('button', {
				name: /download report/i
			})
		);

		jest.runAllTimers();

		fireEvent.click(queryByTestId('submit'));

		await waitFor(() => {
			expect(addAlert).toHaveBeenCalledTimes(1);
		});
	});

	it('displays warning alert when csv reached 10,000 entries', async () => {
		// @ts-ignore

		API.csv.fetchCSV.mockReturnValueOnce(Promise.resolve({ok: true}));

		// @ts-ignore

		API.csv.fetchCount.mockReturnValueOnce(Promise.resolve(11000));

		const {getByRole, queryByTestId} = render(<WrapperComponent />);

		fireEvent.click(
			getByRole('button', {
				name: /download report/i
			})
		);

		jest.runAllTimers();

		fireEvent.click(queryByTestId('submit'));

		await waitFor(() => {
			expect(addAlert).toHaveBeenCalledTimes(2);
		});
	});

	it('displays error alert when csv returns any type of errors', async () => {
		// @ts-ignore

		API.csv.fetchCSV.mockReturnValueOnce(Promise.resolve({ok: false}));

		const {getByRole, queryByTestId} = render(<WrapperComponent />);

		fireEvent.click(
			getByRole('button', {
				name: /download report/i
			})
		);

		jest.runAllTimers();

		fireEvent.click(queryByTestId('submit'));

		await waitFor(() => {
			expect(addAlert).toHaveBeenCalledWith({
				alertType: Alert.Types.Error,
				message: Liferay.Language.get(
					'it-was-not-possible-to-generate-a-csv-file-at-this-moment.-please-try-again-later'
				)
			});
		});
	});
});
