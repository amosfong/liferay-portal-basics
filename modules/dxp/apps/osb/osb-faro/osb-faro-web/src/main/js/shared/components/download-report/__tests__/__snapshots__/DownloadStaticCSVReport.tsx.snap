// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadStaticCSVReport renders component 1`] = `
<div>
  <button
    aria-hidden="true"
    class="btn btn-outline-borderless btn-sm btn-outline-secondary"
    data-suppressed="true"
    type="button"
  >
    <svg
      class="lexicon-icon lexicon-icon-download mr-2"
      role="presentation"
    >
      <use
        href="#download"
      />
    </svg>
    Download Reports
  </button>
  <div
    aria-hidden="true"
    class="modal-backdrop fade show"
    data-suppressed="true"
  />
  <div
    class="fade modal d-block show"
  >
    <div
      class="modal-dialog"
    >
      <div
        aria-labelledby="clay-modal-label-1"
        aria-modal="true"
        class="modal-content"
        role="dialog"
        tabindex="-1"
      >
        <form>
          <div
            class="modal-header"
          >
            <h1
              class="modal-title"
              id="clay-modal-label-1"
            >
              Download Reports
            </h1>
            <button
              aria-label="close"
              class="close btn btn-unstyled"
              type="button"
            >
              <svg
                class="lexicon-icon lexicon-icon-times"
                role="presentation"
              >
                <use
                  href="#times"
                />
              </svg>
            </button>
          </div>
          <div
            class="modal-body"
          >
            <p>
              The generated CSV file supports up to 10,000 entries per export and it will respect the current ordering and search results. Please ensure that any desired changes have been successfully applied before downloading the Individuals list.
            </p>
          </div>
          <div
            class="modal-footer"
          >
            <div
              class="modal-item-first"
            />
            <div
              class="modal-item"
            />
            <div
              class="modal-item-last"
            >
              <div
                class="btn-group"
                role="group"
              >
                <div
                  class="btn-group-item"
                >
                  <button
                    class="btn btn-secondary"
                    data-testid="cancel"
                    type="button"
                  >
                    Cancel
                  </button>
                </div>
                <div
                  class="btn-group-item"
                >
                  <button
                    class="btn btn-primary"
                    data-testid="submit"
                    type="submit"
                  >
                    Download
                  </button>
                </div>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
